### PR TITLE
[feature] Added support for uploading organization specific files #2

### LINKED ISSUE #2: [feature] Easily allow to upload organization specific files (YAML and assets)
We need to add a way to easily deploy organization YAML config files and their related static assets (logo, CSS, etc).

Ideally we could allow specifying all the assets in one directory and upload all those files in some way to the `organizations/` directory.

We need to document this, and would be great to have a basic test in the CI build: eg define some static assets of an org and at the end of the CI build ensure these have been uploaded.

### LINKED ISSUE #3: [tests] Add CI build
We need a CI build with basic tests for this project but I am not sure I will have time to prepare this.
If anyone wants to help me out with this I'll gladly help.

The CI setup can be taken from:

- https://github.com/openwisp/ansible-openwisp2
- https://github.com/openwisp/ansible-wireguard-openwisp/

### PR DESCRIPTION
**Pull request by codesankalp**
Used ansible copy to copy the organizations configuration and assets.
Also added molecule tests.
Added CI build.

Closes #2
Closes #3

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+---
+
+name: Ansible OpenWISP2 CI Build
```
Comment (nemesisdesign) [inline .github/workflows/ci.yml:3]: please update the name here

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+jobs:
+  build:
+    name: Build ${{ matrix.distro }}
+    runs-on: ubuntu-20.04
+
+    strategy:
+      fail-fast: false
+      matrix:
+        distro:
+          - ubuntu1804
```
Comment (nemesisdesign) [inline .github/workflows/ci.yml:22]: we can skip this :)

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+          - ubuntu2004
+          - debian10
```
Comment (nemesisdesign) [inline .github/workflows/ci.yml:24]: would be great to test this on debian11, which has been released recently

File: README.md [hunk 4]
```diff
@@ -10,3 +10,10 @@ Required variables:
 
 Custom static content (HTML files, PDF, etc.) can be deployed
 to ``{{ wifi_login_pages_path }}/static/``.
+
+## Deploy organizations configurations and assets
+
+For deploying organization YAML config files and their related static assets (logo, CSS, etc),
+it is mandatory to paste the `organizations` directory with all the configuration and assets of
+the organizations inside `files/` directory. These files will be uploaded to remote during the
```
Comment (nemesisdesign) [inline README.md:18]: is this the files/ directory of the playbook or of this ansible role?

We should not ask the user to edit the files directory of this role, but it's ok if the user can accomplish this by placing the org config in the `files/` dir of their playbook.

File: files/organizations/default/client_assets/facebook.svg [hunk 5]
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" height="32" width="32"><path d="M32 30a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2V2a2 2 0 0 1 2-2h28a2 2 0 0 1 2 2v28z" fill="#3b5998"/><path d="M22 32V20h4l1-5h-5v-2c0-2 1.002-3 3-3h2V5h-4c-3.675 0-6 2.881-6 7v3h-4v5h4v12h5z" fill="#fff"/></svg>
```
Comment (nemesisdesign) [inline files/organizations/default/client_assets/facebook.svg:1]: why are you adding these files? Are these for testing?

File: molecule/resources/requirements.yml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+
+- src: https://github.com/nemesisdesign/Stouts.postfix
+  version: origin/develop
+  name: Stouts.postfix
```
Comment (nemesisdesign) [inline molecule/resources/requirements.yml:5]: this should not be needed

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Great work @codesankalp!

Could you please add instructions on how to run the tests locally?
Like we have in https://github.com/openwisp/ansible-openwisp2#how-to-run-tests.

See also my comments below.

File: README.md (hunk 4 above)
Comment (codesankalp) [reply README.md:18]: Updated README

File: README.md [hunk 7]
```diff
@@ -4,9 +4,79 @@ Ansible role to deploy and manage [openwisp-wifi-login-pages](https://github.com
...
 
 ## Deploying custom static content
 
 Custom static content (HTML files, PDF, etc.) can be deployed
-to ``{{ wifi_login_pages_path }}/static/``.
+to `{{ wifi_login_pages_path }}/static/`.
+
+## Deploy organizations configurations and assets
+
+For deploying organization YAML config files and their related static assets (logo, CSS, etc), proceed with the following tests:
```
Comment (nemesisdesign) [inline README.md:16]: tests > step?

File: README.md [hunk 8]
```diff
@@ -4,9 +4,79 @@ Ansible role to deploy and manage [openwisp-wifi-login-pages](https://github.com
...
+**Step 2**: Create the directory `files`.
+
+```
+mkdir files
+```
+
+**Step 3**: Copy `organizations` directory with all the organizations configuration and assets inside the created `files` directory.
+
+```
+cp -r <path_to_organizations_directory> files
```
Comment (nemesisdesign) [inline README.md:33]: I think here we should make it explicit that the resulting directory structure should be `files/organizations`, I don't think it's 100% clear

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Great work @codesankalp I think the README needs a tiny bit more improvement, see below.

File: tasks/yarn.yml [hunk 9]
```diff
@@ -47,23 +48,47 @@
...
+    # sets executable only to directories
+    mode: u=rwX,g=rX,o=rX
+    recurse: true
+
+- name: copy organizations
+  become: true
+  copy:
+    src: organizations
+    dest: "{{ wifi_login_pages_path }}"
+    mode: 0644
```
Comment (nemesisdesign) [inline tasks/yarn.yml:72]: can you make sure the files are copied with the www user and group please?

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
I was thinking, just to be 100% sure to avoid conflicts with any other role, I think we should switch the directory name from `organizations/` to `owlp_organizations`, what do you think?

See also my comment below.

File: tasks/yarn.yml (hunk 9 above)
Comment (codesankalp) [reply tasks/yarn.yml:74]: Done

File: tasks/yarn.yml [hunk 10]
```diff
@@ -47,23 +48,49 @@
...
 
-- name: set permissions to client config
+- name: change permissions of server
   file:
-    dest: "{{ wifi_login_pages_path }}/client/config.json"
+    dest: "{{ wifi_login_pages_path }}/server"
     owner: "{{ www_user }}"
     group: "{{ www_group }}"
     mode: 0660
+    recurse: true
```
Comment (nemesisdesign) [inline tasks/yarn.yml:96]: why did you change these two tasks?

I only noticed now that the system where I tested this now is showing permission issues.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

File: tasks/yarn.yml (hunk 10 above)
Comment (nemesisdesign) [reply tasks/yarn.yml:96]: This is the script I am using to rebuild the app:

```
#!/bin/bash
set -e
yarn
# everything owned by www-data
chown -R www-data:www-data .
# remove any public permission
chmod -R o-rwx .
# build in production mode as www-data
sudo -u www-data yarn build --mode production
# restart
supervisorctl restart wifi_login_pages
```

I think we should use something similar here:

- there's no need to keep any public permission
- ensure www-data owns all files
- do not touch any other permission

**Review by nemesisdesign — COMMENTED** [review]
Can you please add a way to upload custom i18n files? Eg: `owlp_i18n`?

**Review by nemesisdesign — APPROVED** [review]
Great work @codesankalp, thanks!

I made some changes, I mainly simplifies the permission changes since I believe we can keep it simple and use one set of permissions for all directories, ensuring we always build as www-data.

I have tested my changes and they're working good.

Please double check and let me know if you have any doubts.

**Comment by codesankalp** [comment]
> Please double check and let me know if you have any doubts.

Yes, it's working with this also.
